yasa/hypno.py

```diff
@@ -48,6 +24,826 @@
+        --------
+        Let's assume that we have an hypnogram where sleep = 1 and wake = 0, with one value
+        per minute.
+
+        >>> from yasa import Hypnogram
+        >>> val = 11 * ["W"] + 3 * ["S"] + 2 * ["W"] + 9 * ["S"] + ["W", "W"]
+        >>> hyp = Hypnogram(val, n_stages=2, freq="1min")
+        >>> hyp.find_periods(threshold="0min")
+          values  start  length
+        0   WAKE      0      11
+        1  SLEEP     11       3
+        2   WAKE     14       2
+        3  SLEEP     16       9
+        4   WAKE     25       2
+
+        This gives us the start and duration of each sequence of consecutive values in the
+        hypnogram. For example, the first row tells us that there is a sequence of 11 consecutive
+        WAKE starting at the first index of hypno.
+
+        Now, we may want to keep only periods that are longer than a specific threshold,
+        for example 5 minutes:
+
+        >>> hyp.find_periods(threshold="5min")
+          values  start  length
+        0   WAKE      0      11
+        1  SLEEP     16       9
+
+        Only the two sequences that are longer than 5 minutes (11 minutes and 9 minutes
+        respectively) are kept. Feel free to play around with different values of threshold!
+
+        This function is not limited to binary arrays, e.g. a 5-stages hypnogram at 30-sec
+        resolution:
+
+        >>> val = ["W", "W", "W", "W", "N1"] + 6 * ["N2"] + ["W", "W", "W", "N1", "W", "N1"]
+        >>> hyp = Hypnogram(val)
+        >>> hyp.find_periods(threshold="2min")
+          values  start  length
+        0   WAKE      0       4
+        1     N2      5       6
+
+        Lastly, using ``equal_length=True`` will further divide the periods into segments of the
+        same duration, i.e. the duration defined in ``threshold``:
+
+        >>> hyp.find_periods(threshold="1min", equal_length=True)
+          values  start  length
+        0   WAKE      0       2
+        1   WAKE      2       2
+        2     N2      5       2
+        3     N2      7       2
+        4     N2      9       2
+        5   WAKE     11       2
+
+        Here, the first 2 minutes of consecutive WAKE are divided into 2 periods of exactly 1
+        minute each. Next, the sequence of 6 consecutive 30-sec N2 epochs is further divided
+        into 3 periods of 1 minute each. Lastly, the last value in the sequence of 3
+        consecutive WAKE at the end of the hypnogram is removed to keep only a segment of exactly
+        1 minute. In other words, the remainder of the division of a given segment by the desired
+        duration is discarded.
+        """
+        return hypno_find_periods(
+            self.hypno, self.sampling_frequency, threshold=threshold, equal_length=equal_length
+        )
+
+    def plot_hypnogram(self):
+        """Plot the hypnogram."""
+        # TODO: Add support for 2, 3 and 4-stages hypnogram
+        raise NotImplementedError
+
+    def sleep_statistics(self):
```

Comment:
I could see this function being useful outside of this class, maybe as a utility in some cases. Do you think it should be left out and then just call it here from the class? Or maybe it's cleaner to just have users make a Hypnogram class quickly to use it. Cool with me.

Reply:
I plan to keep the `yasa.sleep_statistics` function intact for now. Long term however I'd like all users to switch to the Hypnogram class method.

Reply:
Right, but then why not just use that function here? Maybe I'm missing a critical difference but at a quick glance it seems like this would just be repeating lots of the code?

Reply:
In this new implementation I'm really taking advantage of the numerous attributes (e.g. `.labels`, `.n_epochs`, `.sampling_frequency`) and methods (`.transition_matrix()`). I guess these could be passed to the `yasa.sleep_statistics` function but I thought it would be just easier to repeat some code rather than risking breaking the existing function. I can check next week how much work it would be to adapt the original function.

Reply:
No, again don't stress over it. I'm sure you're taking the right approach, I was mostly just curious about your thought process (should've just asked that). This clarification makes sense 👍